src/Oscoin/Storage/Block/SQLite/Tx.hs

```diff
@@ -0,0 +1,162 @@
+{-# LANGUAGE QuasiQuotes          #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Store and retrieve transactions.
+module Oscoin.Storage.Block.SQLite.Tx
+    ( StorableTx
+    , TxRow(..)
+    , TxRowDecodeError
+    , getTx
+    , getBlockTxs
+    , storeTxs
+    ) where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Crypto.Blockchain.Block (BlockHash)
+import           Oscoin.Crypto.Hash (Hash, Hashable, fromHashed, hash)
+import           Oscoin.Crypto.Hash.RealWorld ()
+import           Oscoin.Crypto.PubKey
+import           Oscoin.Crypto.PubKey.RealWorld ()
+import           Oscoin.Data.Ledger (AccountId, Balance, Nonce)
+import           Oscoin.Data.OscoinTx
+
+import           Codec.Serialise
+import qualified Data.ByteString.Lazy as LBS
+
+import           Database.SQLite.Simple as Sql (Connection, Only(..))
+import qualified Database.SQLite.Simple as Sql
+import           Database.SQLite.Simple.FromField (FromField)
+import           Database.SQLite.Simple.FromRow (FromRow, field)
+import           Database.SQLite.Simple.Orphans ()
+import           Database.SQLite.Simple.QQ
+import           Database.SQLite.Simple.ToField (ToField, toField)
+
+-- | Constraints required for a @Tx c@ to be storable in SQLite.
+type StorableTx c =
```

Comment:
I know that I am "that guy" and probably a terrible person, but I would be 👎 for creating this type synonym and for exporting it at the top-level. I have resisted very, very hard the temptation to do so for the crypto stuff, and the reason it's simple (_puts on his Blade Runner goggles and interprets the replicant at the end of the movie_): I have seen with my own eyes what happens if you start to leak all these piecemeal type synonyms in type signatures. Soon enough we will create `type Behemoth c = (StorableTx c, Chocobo c, ...)`  and the poor new programmers joining the codebase will have to keep in their head all these layers of indirections to actually get to the constraints. I have been there, it's terrible 😅 

May I propose a middle ground? If you still want to have the conveniency of `StorableTx` _locally for this module_, I am OK with that. The bit I would suggest we refrain from, it's to export it.

Middle ground (of the middle ground): _if_ you need this _only_ in this module _and_ the SQLite module stuff, may I propose you rename it like `InternalStorableTx`? This way, it would be extremely clear this is an internal detail of the SQLite stuff, we other programmers won't be tempted to make it leak into top-level signatures.

What do you think?

Reply:
:thinking: I kind of agree, but basically just copied what was already there: https://github.com/oscoin/oscoin/pull/607/files#diff-b0b113fc3c1a6c626c272326eda78d7cL67 (and also exported) - Let me see how bad it would be not to export it.

Reply:
I think this synonym is mostly used in the tests where we also require `IsCrpyto c`. We can get rid of it if we extend `IsCrypto c` with the constraints listed here. Also see my comment about the `ToField` and `FromField` instances above. We can get rid of them if we use explicit `fieldFromSerial` instead of just `field` in `FromRow (TxRow c)` below.

Reply:
Since this synonym is used a lot from `SQLite.Internal`, which imports `Tx`, I have to export it. But what I'll do is rename the `Tx` module to `Internal.Tx`.

Reply:
b7ddefe